src/bijectors/coupling.jl

```diff
@@ -0,0 +1,273 @@
+one part to transform, one part to map into the parameter-space of the transform applied
+to the first part, and the last part of the vector is not used for anything.
+
+# Examples
+```julia-repl
+julia> using Bijectors: PartitionMask, partition, combine
+
+julia> m = PartitionMask(3, [1], [2]) # <= assumes input-length 3
+PartitionMask{SparseArrays.SparseMatrixCSC{Bool,Int64}}(
+  [1, 1]  =  true, 
+  [2, 1]  =  true, 
+  [3, 1]  =  true)
+
+julia> # Partition into 3 parts; the last part is inferred to be indices `[3, ]` from
+       # the fact that `[1]` and `[2]` does not make up all indices in `1:3`.
+       x1, x2, x3 = partition(m, [1., 2., 3.])
+([1.0], [2.0], [3.0])
+
+julia> # Recombines the partitions into a vector
+       combine(m, x1, x2, x3)
+3-element Array{Float64,1}:
+ 1.0
+ 2.0
+ 3.0
+```
+Note that the underlying `SparseMatrix` is using `Bool` as the element type. We can also
+specify this to be some other type using the `sp_type` keyword:
+```julia-repl
+julia> m = PartitionMask(3, [1], [2]; sp_type = Float32)
+PartitionMask{SparseArrays.SparseMatrixCSC{Float32,Int64}}(
+  [1, 1]  =  1.0, 
+  [2, 1]  =  1.0, 
+  [3, 1]  =  1.0)
+```
+"""
+struct PartitionMask{A}
+    A_1::A
+    A_2::A
+    A_3::A
+
+    # Only make it possible to construct using matrices
+    PartitionMask(A_1::A, A_2::A, A_3::A) where {A <: AbstractMatrix{<:Real}} = new{A}(A_1, A_2, A_3)
+end
+
+function PartitionMask(
+    n::Int,
+    indices_1::AbstractVector{Int},
+    indices_2::AbstractVector{Int},
+    indices_3::AbstractVector{Int};
+    sp_type = Bool
+)
+    A_1 = spzeros(sp_type, n, length(indices_1));
+    A_2 = spzeros(sp_type, n, length(indices_2));
+    A_3 = spzeros(sp_type, n, length(indices_3));
+
+    for (i, idx) in enumerate(indices_1)
+        A_1[idx, i] = one(sp_type)
+    end
+
+    for (i, idx) in enumerate(indices_2)
+        A_2[idx, i] = one(sp_type)
+    end
+
+    for (i, idx) in enumerate(indices_3)
+        A_3[idx, i] = one(sp_type)
+    end
+
+    return PartitionMask(A_1, A_2, A_3)
+end
+
+PartitionMask(
+    n::Int,
+    indices_1::AbstractVector{Int},
+    indices_2::AbstractVector{Int};
+    kwargs...
+) = PartitionMask(n, indices_1, indices_2, nothing; kwargs...)
+
+PartitionMask(
+    n::Int,
+    indices_1::AbstractVector{Int},
+    indices_2::AbstractVector{Int},
+    indices_3::Nothing;
+    kwargs...
+) = PartitionMask(
+    n, indices_1, indices_2, [i for i in 1:n if i ∉ (indices_1 ∪ indices_2)];
+    kwargs...
+)
+
+PartitionMask(
+    n::Int,
+    indices_1::AbstractVector{Int},
+    indices_2::Nothing,
+    indices_3::AbstractVector{Int};
+    kwargs...
+) = PartitionMask(
+    n, indices_1, [i for i in 1:n if i ∉ (indices_1 ∪ indices_3)], indices_3;
+    kwargs...
+)
+
+"""
+    PartitionMask(n::Int, indices)
+
+Assumes you want to _split_ the vector, where `indices` refer to the 
+parts of the vector you want to apply the bijector to.
+"""
+function PartitionMask(n::Int, indices; sp_type = Bool)
```

Comment:
I think it would be more natural to make `sp_type` a type parameter (it's the element type of `A` anyways) and then define
```suggestion
PartitionMask(n::Int, indices) = PartitionMask{Bool}(n, indices)
function PartitionMask{T}(n::Int, indices) where T
```
I'm wondering though why you do not want to hardcode `Bool` here? `Bool` is the number type that affects promotion the least, so in contrast to integers etc (usually) you do not have to be worried about promoting inputs.

Reply:
> I think it would be more natural to make sp_type a type parameter 

Yeah, you're completely right :+1: 

> I'm wondering though why you do not want to hardcode Bool here? 

https://github.com/TuringLang/Bijectors.jl/pull/48#discussion_r376639755 

In addition, it's more generality at essentially zero cost, so to me it seems like something we'd do just in case someone really needs to be some other type?

Reply:
> I think it would be more natural to make sp_type a type parameter

Should we do `PartitionMask{T, A<:AbstractArray{T}}` then?

Reply:
Yes, that's what I had in mind.

Reply:
Done :+1: